src/main/java/io/sqreen/powerwaf/ByteBufferSerializer.java

```diff
@@ -426,38 +452,58 @@ private int left() {
 
     static class PWArgsArrayBuffer {
         private final ByteBuffer buffer;
-        private final int num;
+        private int start;
+        private int num;
+        private final List<PWArgsBuffer> pwArgsBuffers;
 
         static final PWArgsArrayBuffer EMPTY_BUFFER = new PWArgsArrayBuffer();
 
-        PWArgsArrayBuffer(ByteBuffer buffer, int num) {
+        PWArgsArrayBuffer(ByteBuffer buffer, int start, int num) {
             if (num == 0 || buffer == null) {
                 throw new IllegalArgumentException();
             }
-            this.num = num;
             this.buffer = buffer;
+            this.start = start;
+            this.num = num;
+            buffer.limit(start + num * SIZEOF_PWARGS);
+            buffer.position(start);
+            this.pwArgsBuffers = new ArrayList<>(num);
         }
 
         private PWArgsArrayBuffer() {
-            this.num = 0;
             this.buffer = null;
+            this.start = 0;
+            this.num = 0;
+            this.pwArgsBuffers = null;
+        }
+
+        void reset(int start, int num) {
+            this.start = start;
+            this.num = num;
+            this.buffer.limit(start + num * SIZEOF_PWARGS);
+            this.buffer.position(start);
         }
 
         PWArgsBuffer get(int i) {
             if (i < 0 || i >= num) {
                 throw new ArrayIndexOutOfBoundsException();
             }
             assert this.buffer != null;
-            ByteBuffer slice = offsetBuffer(this.buffer, i * SIZEOF_PWARGS);
-            slice.limit(SIZEOF_PWARGS);
-            return new PWArgsBuffer( slice);
+            while (i >= pwArgsBuffers.size()) {
+                ByteBuffer duplicate = this.buffer.duplicate().order(ByteOrder.nativeOrder());
+                pwArgsBuffers.add(new PWArgsBuffer(duplicate, start));
+            }
+            PWArgsBuffer pwArgsBuffer = pwArgsBuffers.get(i);
+            pwArgsBuffer.reset(start + i * SIZEOF_PWARGS);
+            return pwArgsBuffer;
         }
 
         long getAddress() {
             if (buffer == null) {
                 return NULLPTR;
             }
-            return getByteBufferAddress(buffer);
+            long address = getByteBufferAddress(buffer);
+            return address >= 0 ? address + start : address - start;
```

Comment:
I don't get this. The address shouldn't be negative because amd64 doesn't use the top bits, second, even if they were idk why subtracting instead would be appropriate (the pointer is unsigned but represented as unsigned in java, but addition behaves the same bitwise for signed and unsigned numbers)

Reply:
So the `canonical higher half` which indeed uses the high bit, is normally used by the kernel, but why take any chances? As for subtraction, you are completely right. I have no idea why I wrote that. I'll just do a '+'.